### PR TITLE
Make improvements to the WorkStream colored implementation

### LINKED ISSUE #17694: Improve implementation of TaskFlow-based implementation of WorkStream.
As mentioned in #17642, the current implementation of the colored `WorkStream` implementation uses the approach
```
  TaskGraph graph;
  for (color : colors)
    {
       build graph for this color / set up the parallel-for
    }
    execute the graph
```
This creates quite large graphs, and we also have to wait until the whole graph is built for it to execute its first task. A better scheme would break this into smaller portions:
```
  for (color : colors)
    {
       TaskGraph graph;
       build graph for this color / set up the parallel-for
       execute the graph
    }
```
An even better scheme might start to build the graph for the next color while the current color is running, wait for the current color to finish, and start the next color.

@RyanMoulday FYI

### PR DESCRIPTION
**Pull request by RyanMoulday**
This change makes the improvements suggested in #17694. Namely task graphs are now generated and executed per color, eliminating the need for dependency schemes, and reducing memory overhead. Further, graphs for the subsequent color are generated as the previous color's graph executes, which should lead to a performance improvement.

@bangerth 

Fixes #17694.

### CODE REVIEW

File: include/deal.II/base/work_stream.h [hunk 1]
```diff
@@ -1214,108 +1212,93 @@ namespace WorkStream
...
+                  fu.wait();
+                  // Hand the taskflow's ownership to the executor. The taskflow
+                  // variable is now an empty task graph we can safely
+                  // manipulate again in the next cycle. Link the future to this
+                  // new execution.
+                  fu = executor.run(std::move(taskflow));
+                }
+              else
+                {
+                  fu = executor.run(std::move(taskflow));
```
Comment (tjhei) [inline include/deal.II/base/work_stream.h:1297]: This line is duplicated. Can you move it outside the if?

File: include/deal.II/base/work_stream.h [hunk 2]
```diff
@@ -1190,12 +1189,11 @@ namespace WorkStream
...
-        // tasks which are dependent on all tasks in the same "color" to be
-        // completed before allowing any tasks in the next "color" to be worked
-        // on.
-        tf::Task last_color;
-        tf::Task next_color;
+        tf::Taskflow taskflow;
+
+        // Create a "future" object which eventually contains the execution
+        // result of a taskflow graph and can be used to yield execution
+        tf::Future<void> fu;
```
Comment (tjhei) [inline include/deal.II/base/work_stream.h:1196]: No reason to use a variable name that short.

**Review by tjhei — APPROVED** [review]

File: include/deal.II/base/work_stream.h [hunk 3]
```diff
@@ -1214,108 +1212,83 @@ namespace WorkStream
...
+                // Wait for the previous color to finish executing
+                execution_future.wait();
+              execution_future = executor.run(std::move(taskflow));
             }
-        // Now we run all the tasks in the task graph. They will be run in
-        // parallel and are eligible to run when their dependencies established
-        // above are met.
-        executor.run(taskflow).wait();
+        // Wait for our final execution to finish
+        execution_future.wait();
```
Comment (bangerth) [inline include/deal.II/base/work_stream.h:1291]: I think you have to guard this with `if (number of colors > 0)`.

**Review by bangerth — COMMENTED** [review]
Yes, nice -- this is exactly what I had in mind!

I have only one comment that takes care of the case where the number of colors happens to be zero.

**Comment by RyanMoulday** [comment]
Yes, you are correct. Good catch. I tested out of curiosity and indeed calling .wait() on an empty tf::Future throws an error. It should be fixed now.

**Comment by bangerth** [comment]
You know, this sort of thing would make for a good test! I would definitely approve a pull request (perhaps separate from this one) in which you add a test that checks input with zero colors!

**Review by bangerth — APPROVED** [review]

**Pull request merged**